File: shopping-cart/shopping-cart-scala/shopping-cart/src/main/scala/com/example/shoppingcart/impl/ShoppingCartEntity.scala
package com.example.shoppingcart.impl

import akka.Done
import akka.actor.typed.ActorRef
import akka.actor.typed.ActorSystem
import akka.actor.typed.Behavior
import akka.cluster.sharding.typed.scaladsl._
import akka.persistence.journal.Tagged
import akka.persistence.typed.ExpectingReply
import akka.persistence.typed.PersistenceId
import akka.persistence.typed.scaladsl.Effect
import akka.persistence.typed.scaladsl.EventSourcedBehavior
import akka.persistence.typed.scaladsl.ReplyEffect
import com.lightbend.lagom.scaladsl.persistence.AggregateEvent
import com.lightbend.lagom.scaladsl.persistence.AggregateEventShards
import com.lightbend.lagom.scaladsl.persistence.AggregateEventTag
import com.lightbend.lagom.scaladsl.playjson.JsonSerializer
import com.lightbend.lagom.scaladsl.playjson.JsonSerializerRegistry
import play.api.libs.json.Format
import play.api.libs.json._

import scala.collection.immutable.Seq
import java.time.Instant

/**
 * The current state held by the persistent entity.
 */
case class ShoppingCartState(items: Map[String, Int], checkedOut: Boolean) {

  def applyCommand(cmd: ShoppingCartCommand[_]): ReplyEffect[ShoppingCartEvent, ShoppingCartState] =
    cmd match {
      case x: UpdateItem => onUpdateItem(x)
      case x: Checkout   => onCheckout(x)
      case x: Get        => onReadState(x)
    }

  def onUpdateItem(cmd: UpdateItem): ReplyEffect[ShoppingCartEvent, ShoppingCartState] =
    cmd match {
      case UpdateItem(_, qty, _) if qty < 0 =>
        Effect.reply(cmd)(Rejected("Quantity must be greater than zero"))

      case UpdateItem(productId, 0, _) if !items.contains(productId) =>
        Effect.reply(cmd)(Rejected("Cannot delete item that is not already in cart"))

      case UpdateItem(productId, quantity, _) =>
        Effect
          .persist(ItemUpdated(productId, quantity, Instant.now()))
          .thenReply(cmd) { _ =>
            Accepted
          }
    }

  def onCheckout(cmd: Checkout): ReplyEffect[ShoppingCartEvent, ShoppingCartState] =
    if (items.isEmpty)
      Effect.reply(cmd)(Rejected("Cannot checkout empty cart"))
    else
      Effect
        .persist(CheckedOut(Instant.now()))
        .thenReply(cmd) { _ =>
          Accepted
        }

  def onReadState(cmd: Get): ReplyEffect[ShoppingCartEvent, ShoppingCartState] =
    Effect.reply(cmd)(CurrentState(this))

  def applyEvent(evt: ShoppingCartEvent): ShoppingCartState = {
    evt match {
      case ItemUpdated(productId, quantity, _) => updateItem(productId, quantity)
      case CheckedOut(_)                       => checkout
    }
  }

  private def updateItem(productId: String, quantity: Int) = {
    quantity match {
      case 0 => copy(items = items - productId)
      case _ => copy(items = items + (productId -> quantity))
    }
  }

  private def checkout = copy(checkedOut = true)
}

object ShoppingCartState {

  def empty: ShoppingCartState = ShoppingCartState(Map.empty, checkedOut = false)

  val typeKey = EntityTypeKey[ShoppingCartCommand[_]]("ShoppingCartEntity")

  def behavior(entityContext: EntityContext): Behavior[ShoppingCartCommand[_]] = {

    val persistenceId = typeKey.persistenceIdFrom(entityContext.entityId)

    EventSourcedBehavior
      .withEnforcedReplies[ShoppingCartCommand[_], ShoppingCartEvent, ShoppingCartState](
        persistenceId = persistenceId,
        emptyState = ShoppingCartState.empty,
        commandHandler = (cart, cmd) => cart.applyCommand(cmd),
        eventHandler = (cart, evt) => cart.applyEvent(evt)
      )
      .withTagger(LagomTaggerAdapter(entityContext, ShoppingCartEvent.Tag))

  }

  /**
   * Format for the hello state.
   *
   * Persisted entities get snapshotted every configured number of events. This
   * means the state gets stored to the database, so that when the entity gets
   * loaded, you don't need to replay all the events, just the ones since the
   * snapshot. Hence, a JSON format needs to be declared so that it can be
   * serialized and deserialized when storing to and from the database.
   */
  implicit val format: Format[ShoppingCartState] = Json.format
}

/**
 * This interface defines all the events that the ShoppingCartEntity supports.
 */
sealed trait ShoppingCartEvent extends AggregateEvent[ShoppingCartEvent] {
  def aggregateTag = ShoppingCartEvent.Tag
}

object ShoppingCartEvent {
  val Tag = AggregateEventTag.sharded[ShoppingCartEvent](numShards = 10)
}

/**
 * An event that represents a item updated event.
 */
final case class ItemUpdated(productId: String, quantity: Int, eventTime: Instant) extends ShoppingCartEvent

object ItemUpdated {

  /**
   * Format for the item updated event.
   *
   * Events get stored and loaded from the database, hence a JSON format
   * needs to be declared so that they can be serialized and deserialized.
   */
  implicit val format: Format[ItemUpdated] = Json.format
}

final case class CheckedOut(eventTime: Instant) extends ShoppingCartEvent

object CheckedOut {

  /**
   * Format for the checked out event.
   *
   * Events get stored and loaded from the database, hence a JSON format
   * needs to be declared so that they can be serialized and deserialized.
   */
  implicit val format: Format[CheckedOut] = Json.format
}

sealed trait ShoppingCartReply

object ShoppingCartReply {
  implicit val format: Format[ShoppingCartReply] =
    new Format[ShoppingCartReply] {

      override def reads(json: JsValue): JsResult[ShoppingCartReply] = {
        if ((json \ "state").isDefined)
          Json.fromJson[CurrentState](json)
        else
          Json.fromJson[Confirmation](json)
      }

      override def writes(o: ShoppingCartReply): JsValue = {
        o match {
          case conf: Confirmation  => Json.toJson(conf)
          case state: CurrentState => Json.toJson(state)
        }
      }
    }
}

sealed trait Confirmation extends ShoppingCartReply

case object Confirmation {
  implicit val format: Format[Confirmation] = new Format[Confirmation] {
    override def reads(json: JsValue): JsResult[Confirmation] = {
      if ((json \ "reason").isDefined)
        Json.fromJson[Rejected](json)
      else
        Json.fromJson[Accepted](json)
    }

    override def writes(o: Confirmation): JsValue = {
      o match {
        case acc: Accepted => Json.toJson(acc)
        case rej: Rejected => Json.toJson(rej)
      }
    }
  }
}

sealed trait Accepted extends Confirmation

case object Accepted extends Accepted {
  implicit val format: Format[Accepted] = Format(
    Reads(_ => JsSuccess(Accepted)),
    Writes(_ => Json.obj())
  )
}

case class Rejected(reason: String) extends Confirmation

object Rejected {
  implicit val format: Format[Rejected] = Json.format
}

final case class CurrentState(state: ShoppingCartState) extends ShoppingCartReply

object CurrentState {
  implicit val format: Format[CurrentState] = Json.format
}

/**
 * This is a marker trait for shopping cart commands.
 * We will serialize them using Akka's Jackson support that is able to deal with the replyTo field.
 * (see application.conf)
 */
trait ShoppingCartCommandSerializable

/**
 * This interface defines all the commands that the ShoppingCartEntity supports.
 */
sealed trait ShoppingCartCommand[R <: ShoppingCartReply] extends ExpectingReply[R] with ShoppingCartCommandSerializable

/**
 * A command to update an item.
 *
 * It has a reply type of [[Done]], which is sent back to the caller
 * when all the events emitted by this command are successfully persisted.
 */
case class UpdateItem(productId: String, quantity: Int, replyTo: ActorRef[Confirmation])
    extends ShoppingCartCommand[Confirmation]

/**
 * A command to get the current state of the shopping cart.
 *
 * The reply type is the ShoppingCart, and will contain the message to say to that
 * person.
 */
case class Get(replyTo: ActorRef[CurrentState]) extends ShoppingCartCommand[CurrentState]

/**
 * A command to checkout the shopping cart.
 *
 * The reply type is the Done, which will be returned when the events have been
 * emitted.
 */
case class Checkout(replyTo: ActorRef[Confirmation]) extends ShoppingCartCommand[Confirmation]

/**
 * Akka serialization, used by both persistence and remoting, needs to have
 * serializers registered for every type serialized or deserialized. While it's
 * possible to use any serializer you want for Akka messages, out of the box
 * Lagom provides support for JSON, via this registry abstraction.
 *
 * The serializers are registered here, and then provided to Lagom in the
 * application loader.
 */
object ShoppingCartSerializerRegistry extends JsonSerializerRegistry {
  override def serializers: Seq[JsonSerializer[_]] = Seq(
    // state and events can use play-json, but commands should use jackson because of ActorRef[T] (see application.conf)
    JsonSerializer[ShoppingCartState],
    JsonSerializer[ItemUpdated],
    JsonSerializer[CheckedOut],
    // the replies use play-json as well
    JsonSerializer[ShoppingCartReply],
    JsonSerializer[CurrentState],
    JsonSerializer[Confirmation],
    JsonSerializer[Accepted],
    JsonSerializer[Rejected]
  )
}

File: shopping-cart/shopping-cart-java/shopping-cart/src/main/java/com/example/shoppingcart/impl/ShoppingCartEntity.java
package com.example.shoppingcart.impl;

import akka.actor.typed.ActorRef;
import akka.cluster.sharding.typed.javadsl.EntityContext;
import akka.cluster.sharding.typed.javadsl.EntityTypeKey;
import akka.persistence.typed.PersistenceId;
import akka.persistence.typed.javadsl.*;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.google.common.base.Preconditions;
import com.lightbend.lagom.javadsl.persistence.AggregateEvent;
import com.lightbend.lagom.javadsl.persistence.AggregateEventShards;
import com.lightbend.lagom.javadsl.persistence.AggregateEventTag;
import com.lightbend.lagom.javadsl.persistence.AggregateEventTagger;
import com.lightbend.lagom.javadsl.persistence.AkkaTaggerAdapter;
import com.lightbend.lagom.serialization.CompressedJsonable;
import com.lightbend.lagom.serialization.Jsonable;
import lombok.Value;
import org.pcollections.HashTreePMap;
import org.pcollections.PMap;

import java.time.Instant;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;

public class ShoppingCartEntity extends EventSourcedBehaviorWithEnforcedReplies<ShoppingCartEntity.Command, ShoppingCartEntity.Event, ShoppingCartEntity.ShoppingCart> {

    private final String cartId;
    
    private final Function <Event, Set<String>> tagger;
    
    static EntityTypeKey<Command> ENTITY_TYPE_KEY = EntityTypeKey.create(Command.class, "ShoppingCart");
    
    private ShoppingCartEntity(EntityContext<Command> entityContext) {
        // PersistenceId needs a typeHint (or namespace) and entityId, we take then from the EntityContext
        super(PersistenceId.of(entityContext.getEntityTypeKey().name(), entityContext.getEntityId()));
        // we keep a copy of cartId because it's used in the events
        this.cartId = entityContext.getEntityId();
        // tagger is constructed from adapter and needs EntityContext
        this.tagger = AkkaTaggerAdapter.fromLagom(entityContext, Event.TAG);
    }

    static ShoppingCartEntity create(EntityContext<Command> entityContext) {
        return new ShoppingCartEntity(entityContext);
    }

    //
    // SHOPPING CART COMMANDS
    //
    interface Command<R> extends Jsonable {}

    @Value
    @JsonDeserialize
    static final class AddItem implements Command<Confirmation>, CompressedJsonable {
        public final String itemId;
        public final int quantity;
        public final ActorRef<Confirmation> replyTo;

        @JsonCreator
        AddItem(String itemId, int quantity, ActorRef<Confirmation> replyTo) {
            this.itemId = Preconditions.checkNotNull(itemId, "itemId");
            this.quantity = quantity;
            this.replyTo = replyTo;
        }
    }

    @Value
    @JsonDeserialize
    static final class RemoveItem implements Command<Confirmation> {
        public final String itemId;
        public final ActorRef<Confirmation> replyTo;

        @JsonCreator
        RemoveItem(String itemId, ActorRef<Confirmation> replyTo) {
            this.itemId = Preconditions.checkNotNull(itemId, "itemId");
            this.replyTo = replyTo;
        }
    }

    @Value
    @JsonDeserialize
    static final class AdjustItemQuantity implements Command<Confirmation>, CompressedJsonable {
        public final String itemId;
        public final int quantity;
        public final ActorRef<Confirmation> replyTo;

        @JsonCreator
        AdjustItemQuantity(String itemId, int quantity, ActorRef<Confirmation> replyTo) {
            this.itemId = Preconditions.checkNotNull(itemId, "itemId");
            this.quantity = quantity;
            this.replyTo = replyTo;
        }
    }

    static final class Get implements Command<Summary> {
        private final ActorRef<Summary> replyTo;

        @JsonCreator
        Get(ActorRef<Summary> replyTo) {
            this.replyTo = replyTo;
        }
    }

    static final class Checkout implements Command<Confirmation> {
        private final ActorRef<Confirmation> replyTo;

        @JsonCreator
        Checkout(ActorRef<Confirmation> replyTo) {
            this.replyTo = replyTo;
        }
    }

    //
    // SHOPPING CART REPLIES
    //
    interface Reply extends Jsonable {}

    interface Confirmation extends Reply {}

    @Value
    @JsonDeserialize
    static final class Summary implements Reply {

        public final Map<String, Integer> items;
        public final boolean checkedOut;
        public final Optional<Instant> checkoutDate;

        @JsonCreator
        Summary(Map<String, Integer> items, boolean checkedOut, Optional<Instant> checkoutDate) {
            this.items = items;
            this.checkedOut = checkedOut;
            this.checkoutDate = checkoutDate;
        }
    }

    @Value
    @JsonDeserialize
    static final class Accepted implements Confirmation {
        public final Summary summary;

        @JsonCreator
        Accepted(Summary summary) {
            this.summary = summary;
        }
    }

    @Value
    @JsonDeserialize
    static final class Rejected implements Confirmation {
        public final String reason;

        @JsonCreator
        Rejected(String reason) {
            this.reason = reason;
        }
    }

    //
    // SHOPPING CART EVENTS
    //
    public interface Event extends Jsonable, AggregateEvent<Event> {
        /**
         * The tag for shopping cart events, used for consuming the Journal event stream later.
         */
        AggregateEventShards<Event> TAG = AggregateEventTag.sharded(Event.class, 10);

        @Override
        default AggregateEventTagger<Event> aggregateTag() {
            return TAG;
        }
    }

    @Value
    @JsonDeserialize
    static final class ItemAdded implements Event {
        public final String shoppingCartId;
        public final String itemId;
        public final int quantity;
        public final Instant eventTime;

        @JsonCreator
        ItemAdded(String shoppingCartId, String itemId, int quantity, Instant eventTime) {
            this.shoppingCartId = Preconditions.checkNotNull(shoppingCartId, "shoppingCartId");
            this.itemId = Preconditions.checkNotNull(itemId, "itemId");
            this.quantity = quantity;
            this.eventTime = eventTime;
        }
    }

    @Value
    @JsonDeserialize
    static final class ItemRemoved implements Event {
        public final String shoppingCartId;
        public final String itemId;
        public final Instant eventTime;

        @JsonCreator
        ItemRemoved(String shoppingCartId, String itemId, Instant eventTime) {
            this.shoppingCartId = Preconditions.checkNotNull(shoppingCartId, "shoppingCartId");
            this.itemId = Preconditions.checkNotNull(itemId, "itemId");
            this.eventTime = eventTime;
        }
    }

    @Value
    @JsonDeserialize
    static final class ItemQuantityAdjusted implements Event {
        public final String shoppingCartId;
        public final String itemId;
        public final int quantity;
        public final Instant eventTime;

        @JsonCreator
        ItemQuantityAdjusted(String shoppingCartId, String itemId, int quantity, Instant eventTime) {
            this.shoppingCartId = Preconditions.checkNotNull(shoppingCartId, "shoppingCartId");
            this.itemId = Preconditions.checkNotNull(itemId, "itemId");
            this.quantity = quantity;
            this.eventTime = eventTime;
        }
    }

    @Value
    @JsonDeserialize
    static final class CheckedOut implements Event {

        public final String shoppingCartId;
        public final Instant eventTime;

        @JsonCreator
        CheckedOut(String shoppingCartId, Instant eventTime) {
            this.shoppingCartId = Preconditions.checkNotNull(shoppingCartId, "shoppingCartId");
            this.eventTime = eventTime;
        }
    }

    //
    // SHOPPING CART STATE
    //
    @Value
    @JsonDeserialize
    static final class ShoppingCart implements CompressedJsonable {

        public final PMap<String, Integer> items;
        public final Optional<Instant> checkoutDate;

        @JsonCreator
        ShoppingCart(PMap<String, Integer> items, Instant checkoutDate) {
            this.items = Preconditions.checkNotNull(items, "items");
            this.checkoutDate = Optional.ofNullable(checkoutDate);
        }

        ShoppingCart removeItem(String itemId) {
            PMap<String, Integer> newItems = items.minus(itemId);
            return new ShoppingCart(newItems, null);
        }

        ShoppingCart updateItem(String itemId, int quantity) {
            PMap<String, Integer> newItems = items.plus(itemId, quantity);
            return new ShoppingCart(newItems, null);
        }

        boolean isEmpty() {
            return items.isEmpty();
        }

        boolean hasItem(String itemId) {
            return items.containsKey(itemId);
        }

        ShoppingCart checkout(Instant when) {
            return new ShoppingCart(items, when);
        }

        boolean isOpen() {
            return !this.isCheckedOut();
        }

        boolean isCheckedOut() {
            return this.checkoutDate.isPresent();
        }

        public static final ShoppingCart EMPTY = new ShoppingCart(HashTreePMap.empty(), null);
    }

    @Override
    public ShoppingCart emptyState() {
        return ShoppingCart.EMPTY;
    }

    @Override
    public RetentionCriteria retentionCriteria() {
       return RetentionCriteria.snapshotEvery(100, 2);
    }

    @Override
    public Set<String> tagsFor(Event event) {
        return tagger.apply(event);
    }

    @Override
    public CommandHandlerWithReply<Command, Event, ShoppingCart> commandHandler() {
        CommandHandlerWithReplyBuilder<Command, Event, ShoppingCart> builder = newCommandHandlerWithReplyBuilder();
        builder.forState(ShoppingCart::isOpen)
                .onCommand(AddItem.class, this::onAddItem)
                .onCommand(RemoveItem.class, this::onRemoveItem)
                .onCommand(AdjustItemQuantity.class, this::onAdjustItemQuantity)
                .onCommand(Checkout.class, this::onCheckout);

        builder.forState(ShoppingCart::isCheckedOut)
                .onCommand(AddItem.class, cmd -> Effect().reply(cmd.replyTo, new Rejected("Cannot add an item to a checked-out cart")))
                .onCommand(RemoveItem.class, cmd -> Effect().reply(cmd.replyTo, new Rejected("Cannot remove an item to a checked-out cart")))
                .onCommand(AdjustItemQuantity.class, cmd -> Effect().reply(cmd.replyTo, new Rejected("Cannot adjust item quantity in a checked-out cart")))
                .onCommand(Checkout.class, cmd -> Effect().reply(cmd.replyTo, new Rejected("Cannot checkout a checked-out cart")));

        builder.forAnyState().onCommand(Get.class, this::onGet);
        return builder.build();
    }

    private ReplyEffect<Event, ShoppingCart> onAddItem(ShoppingCart shoppingCart, AddItem cmd) {
        if (shoppingCart.hasItem(cmd.getItemId())) {
            return Effect().reply(cmd.replyTo, new Rejected("Item was already added to this shopping cart"));
        } else if (cmd.getQuantity() <= 0) {
            return Effect().reply(cmd.replyTo, new Rejected("Quantity must be greater than zero"));
        } else {
            return Effect()
                    .persist(new ItemAdded(cartId, cmd.getItemId(), cmd.getQuantity(), Instant.now()))
                    .thenReply(cmd.replyTo, s -> new Accepted(toSummary(s)));
        }
    }

    private ReplyEffect<Event, ShoppingCart> onRemoveItem(ShoppingCart shoppingCart, RemoveItem cmd) {
        if (shoppingCart.hasItem(cmd.getItemId())) {
            return Effect()
                    .persist(new ItemRemoved(cartId, cmd.getItemId(), Instant.now()))
                    .thenReply(cmd.replyTo, updatedShoppingCart -> new Accepted(toSummary(updatedShoppingCart)));
        } else {
            // Remove is idempotent, so we can just return the summary here
            return Effect().reply(cmd.replyTo, new Accepted(toSummary(shoppingCart)));
        }
    }

    private ReplyEffect<Event, ShoppingCart> onAdjustItemQuantity(ShoppingCart shoppingCart, AdjustItemQuantity cmd) {
        if (cmd.getQuantity() <= 0) {
            return Effect().reply(cmd.replyTo, new Rejected("Quantity must be greater than zero"));
        } else if (shoppingCart.hasItem(cmd.getItemId())) {
            return Effect()
                    .persist(new ItemQuantityAdjusted(cartId, cmd.getItemId(), cmd.getQuantity(), Instant.now()))
                    .thenReply(cmd.replyTo, s -> new Accepted(toSummary(s)));
        } else {
            return Effect().reply(cmd.replyTo, new Rejected("Item not found in shopping cart"));
        }
    }

    private ReplyEffect<Event, ShoppingCart> onGet(ShoppingCart shoppingCart, Get cmd) {
        return Effect().reply(cmd.replyTo, toSummary(shoppingCart));
    }

    private ReplyEffect<Event, ShoppingCart> onCheckout(ShoppingCart shoppingCart, Checkout cmd) {
        if (shoppingCart.isEmpty()) {
            return Effect().reply(cmd.replyTo, new Rejected("Cannot checkout empty shopping cart"));
        } else {
            return Effect().persist(new CheckedOut(cartId, Instant.now())).thenReply(cmd.replyTo, s -> new Accepted(toSummary(s)));
        }
    }

    @Override
    public EventHandler<ShoppingCart, Event> eventHandler() {
        return newEventHandlerBuilder()
                .forAnyState()
                .onEvent(ItemAdded.class, (shoppingCart, evt) -> shoppingCart.updateItem(evt.getItemId(), evt.getQuantity()))
                .onEvent(ItemRemoved.class, (shoppingCart, evt) -> shoppingCart.removeItem(evt.getItemId()))
                .onEvent(ItemQuantityAdjusted.class, (shoppingCart, evt) -> shoppingCart.updateItem(evt.getItemId(), evt.getQuantity()))
                .onEvent(CheckedOut.class, (shoppingCart, evt) -> shoppingCart.checkout(evt.getEventTime()))
                .build();
    }

    private Summary toSummary(ShoppingCart shoppingCart) {
        return new Summary(shoppingCart.getItems(), shoppingCart.isCheckedOut(), shoppingCart.getCheckoutDate());
    }
}
